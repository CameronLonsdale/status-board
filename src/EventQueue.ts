/**
 * Event Queue
 *
 * @param io
 * @constructor
 */
export default class EventQueue {
  private io: any;
  private latestEvents: any;

  constructor(io: any) {
    this.io = io;
    this.latestEvents = {};

    io.on('connection', (socket: any) => {
      socket.on('resend', (data: any) => {
        if (this.latestEvents[data]) {
          socket.emit(data, this.latestEvents[data]);
        }
      });

      // broadcast logs
      socket.on('log', (data: any) => {
        socket.broadcast.emit('client', data);
      });
    });
  }

  /**
   * Send widget data to clients
   * @param id
   * @param data
   */
  public send(id: any, data: any) {
    this.latestEvents[id] = data;
    this.io.emit(id, data); // emit to widget
    this.io.emit('client', { data, widgetId: id }); // emit to logger
  }
}
