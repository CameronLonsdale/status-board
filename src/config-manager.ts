import * as bug from 'debug';
import * as fs from 'fs';
import * as path from 'path';
import * as extend from 'xtend';
import { logger } from './logger';

const debug = bug('config-manager');

function readConfigIfExists(fileName: string) {
  let configFileName = fileName;

  if (!path.extname(configFileName)) {
    configFileName = configFileName + '.js';
  }

  if (fs.existsSync(configFileName)) {
    return require(configFileName);
  }

  return {};
}

function readEnv(configFileName: string) {
  const key = 'ATLASBOARD_CONFIG_' + configFileName;

  debug('ENV key', key);

  const environmentKey = process.env[key];

  if (environmentKey) {

    debug('ENV configuration found for', key);

    try {
      const configValue = JSON.parse(environmentKey);

      if (typeof configValue === 'object') {
        return JSON.parse(environmentKey);
      }

      logger.error(`
        ENV configuration key ${key} could not be serialized into an object: ${environmentKey}
      `);

    } catch (error) {
      throw Error(`
        ENV configuration key ${key} contains invalid JSON: ${environmentKey}
        Error: ${error}
      `);
    }
  }
}

export default function (configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
