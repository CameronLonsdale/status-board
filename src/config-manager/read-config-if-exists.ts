import * as fs from 'fs';
import * as path from 'path';

export function readConfigIfExists(fileName: string) {
  let configFileName = fileName;

  if (!path.extname(configFileName)) {
    configFileName = configFileName + '.js';
  }

  if (fs.existsSync(configFileName)) {
    return require(configFileName);
  }

  return {};
}
