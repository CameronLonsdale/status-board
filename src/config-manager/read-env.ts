import * as bug from 'debug';

const debug = bug('config-manager');

export function readEnv(configFileName: string) {
  const key = 'ATLASBOARD_CONFIG_' + configFileName;

  debug('ENV key', key);

  const environmentKey = process.env[key];

  if (environmentKey) {

    debug('ENV configuration found for', key);

    try {
      return JSON.parse(environmentKey);
    } catch (error) {
      throw Error(`
        ENV configuration key ${key} contains invalid JSON: ${environmentKey}
        Error: ${error}
      `);
    }
  }
}
