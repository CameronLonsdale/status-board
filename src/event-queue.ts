/**
 * Event Queue
 *
 * @param io
 * @constructor
 */
export default function eventQueue(io: any) {
  this.io = io;
  this.latestEvents = {};

  io.on('connection', (socket: any) => {
    socket.on('resend', (data: any) => {
      if (this.latestEvents[data]) {
        socket.emit(data, this.latestEvents[data]);
      }
    });

    // broadcast logs
    socket.on('log', (data: any) => {
      socket.broadcast.emit('client', data);
    });
  });

  return {
    /**
     * Send widget data to clients
     * @param id
     * @param data
     */
    send: (id: any, data: any) => {
      this.latestEvents[id] = data;
      this.io.emit(id, data); // emit to widget
      this.io.emit('client', { data, widgetId: id }); // emit to logger
    },
  };
}
