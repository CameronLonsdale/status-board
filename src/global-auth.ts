import * as fs from 'fs';
import * as traverse from 'traverse';

import logger from './logger';

const ENV_VAR_REGEX = /\$\{([^}]+)\}/;

export function globalAuth(file: any) {
  let auth = {};

  try {
    auth = JSON.parse(fs.readFileSync(file).toString());
  } catch (e) {
    if (e.code === 'ENOENT') {
      logger().warn(`Authentication file not found in ${file}.`);
      logger().warn('You may want to create your own.');
      logger().warn('You can also define the place where the credential file will be located by ' +
        'editing the auth file configuration property \'authenticationFilePath\'');
    } else {
      logger().error('Error reading ' + file + '. It may contain invalid json format');
    }
    return auth;
  }

  try {
    traverse(auth).forEach(function (val) {
      if ('string' === typeof val) {
        const match = ENV_VAR_REGEX.exec(val);
        let modified;
        let newVal;
        if (match !== null) {
          const envName = match[1];
          let envVal = process.env[envName];
          if (envVal === undefined) {
            // tslint:disable-next-line max-line-length
            logger().warn(`Authentication file referenced var \${${envName}}, which was not present in environment.`);
            envVal = '';
          }

          newVal = val.substring(
            0,
            match.index,
          ) + envVal + val.substring(
            match.index + match[0].length,
          );
          modified = true;
        }

        if (modified) {
          this.update(newVal);
        } else {
          this.update(val);
        }
      }
    });
  } catch (error) {
    throw Error(`
      Error parsing the auth file ${file} with env variables.
      Error: ${error}
    `);
  }

  return auth;
}
