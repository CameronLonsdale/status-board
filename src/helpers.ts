import * as fs from 'fs';

export function readJSONFile(dashboardPath: any, cb: any) {
  fs.readFile(dashboardPath, (error: any, data: any) => {
    if (error) {
      return cb(error);
    }

    try {
      cb(null, JSON.parse(data));
    } catch (error) {
      cb(error);
    }
  });
}

export function noop() {
  // No operation performed.
}
