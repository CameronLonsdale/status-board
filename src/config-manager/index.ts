import * as path from 'path';
import * as extend from 'xtend';
import { readConfigIfExists } from './read-config-if-exists';
import { readEnv } from './read-env';

export function configManager(configFileName: any) {
  const localConfigFilePath = path.join(process.cwd(), 'config', configFileName);
  const atlasboardConfigFilePath = path.join(__dirname, '../../config/', configFileName);

  return extend(
    readConfigIfExists(atlasboardConfigFilePath),
    readConfigIfExists(localConfigFilePath),
    readEnv(configFileName),
  );
}
