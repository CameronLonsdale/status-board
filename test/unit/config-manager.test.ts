import { configManager } from '../../src/config-manager';

describe('Config Manager', () => {
  test('Should match snapshot', () => {
    expect(configManager).toMatchSnapshot();
  });
});
