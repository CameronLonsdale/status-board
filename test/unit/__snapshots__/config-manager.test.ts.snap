// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Manager Should match snapshot 1`] = `[Function]`;
